rust/automerge-wasm/src/lib.rs

```diff
@@ -786,6 +787,212 @@ impl Automerge {
         let hash = self.doc.empty_change(options);
         JsValue::from_str(&hex::encode(hash))
     }
+
+    pub fn mark(
+        &mut self,
+        obj: JsValue,
+        range: JsValue,
+        name: JsValue,
+        value: JsValue,
+        datatype: JsValue,
+    ) -> Result<(), JsValue> {
+        let (obj, _) = self.import(obj)?;
+        let re = Regex::new(r"([\[\(])(\d+)\.\.(\d+)([\)\]])").unwrap();
+        let range = range.as_string().ok_or("range must be a string")?;
+        let cap = re.captures_iter(&range).next().ok_or(format!("(range={}) range must be in the form of (start..end] or [start..end) etc... () for sticky, [] for normal",range))?;
+        let start: usize = cap[2].parse().map_err(|_| to_js_err("invalid start"))?;
+        let end: usize = cap[3].parse().map_err(|_| to_js_err("invalid end"))?;
+        let start_sticky = &cap[1] == "(";
+        let end_sticky = &cap[4] == ")";
+        let name = name
+            .as_string()
+            .ok_or("invalid mark name")
+            .map_err(to_js_err)?;
+        let value = self
+            .import_scalar(&value, &datatype.as_string())
+            .ok_or_else(|| to_js_err("invalid value"))?;
+        self.doc
+            .mark(
+                &obj,
+                am::marks::Mark::new(name, value, start, end, start_sticky, end_sticky),
+            )
+            .map_err(to_js_err)?;
+        Ok(())
+    }
+
+    pub fn unmark(&mut self, obj: JsValue, mark: JsValue) -> Result<(), JsValue> {
+        let (obj, _) = self.import(obj)?;
+        let mark = self.import_obj(mark)?;
+        self.doc.unmark(&obj, &mark).map_err(to_js_err)?;
+        Ok(())
+    }
+
+    pub fn spans(&mut self, obj: JsValue) -> Result<JsValue, JsValue> {
```

Comment:
:question: I am a little confused about the difference between a span and a mark.

As an example, would bolding a sentence and then unbolding a word in the middle of the sentence require something like the following?

```ts
let list = doc.putObject("_root", "list", "");

// Initially, bold the whole sentence
doc.splice(list, 0, 0, "aaabbbccc");
doc.mark(list, "[0..9]", "bold" , true);

// Then, to unbold the 'bbb' we need to find the correct span in raw_spans, ...
const raw_spans = doc.raw_spans(list);
const the_span = raw_spans[0];
// ... unmark the span/mark ...
doc.unmark(list, the_span.id);
// and then add 2 marks
doc.mark(list, "[0..3]", "bold" , true)
doc.mark(list, "[6..9]", "bold" , true)
```

Rather than alternating terminology, would it be helpful to name the methods to be more mark (or span)-specific? For example:

```ts
let list = doc.putObject("_root", "list", "");

// Initially, bold the whole sentence
doc.splice(list, 0, 0, "aaabbbccc");
const mark = doc.mark(list, "[0..9]", "bold" , true);

// Then, to unbold the 'bbb' we call unmark
doc.unmark(span.id);

// Or, if the client uses the markup() method to render it could be something like this:
const markup = doc.markup(list);
// [
//   [ [ '39@aaabbbccc', 'bold', 'boolean', true ] ],
//   'aaabbbccc',
//   [],
// ]
doc.unmark(markup[0][0][0]);

// and then add 2 marks
doc.mark(list, "[0..3]", "bold" , true);
doc.mark(list, "[6..9]", "bold" , true);
```

Reply:
You are correct - I got marks and spans mixed up when naming those functions.  With the last push I removed `spans()` and added `marks()` and renamed `raw_spans()` to `rawMarks()`.  `rawMarks` is intended for debugging and will likely be removed before this goes live.  Also unsure if `unmark()` will be in the final api.

As for how to unbold a section - you should do it like this.

```
  doc.mark(list, "[0..9]", "bold" , true);
  doc.mark(list, "[3..6]", "bold" , false);
```

This way it'll play nicely with other concurrent marks being made by other users.